File: sdks/typescript/src/observe.ts
/** `observe` higher-order wrapper and `span` helper. */

import { context, Span, SpanStatusCode, trace } from "@opentelemetry/api";

import { Attr, SpanKind, SpanKindValue } from "./semconv";
import { getConfig, getTracer } from "./tracing";

let seq = 0;

function toValue(obj: unknown): string {
  if (typeof obj === "string") return obj;
  try {
    return JSON.stringify(obj);
  } catch {
    return String(obj);
  }
}

function applyCommon(span: Span, kind: SpanKindValue): void {
  span.setAttribute(Attr.SPAN_KIND, kind);
  span.setAttribute(Attr.SEQ, ++seq);
  const config = getConfig();
  if (config?.releaseId) {
    span.setAttribute(Attr.RELEASE_ID, config.releaseId);
  }
}

export function setInput(value: unknown): void {
  const span = trace.getActiveSpan();
  if (span) span.setAttribute(Attr.INPUT_VALUE, toValue(value));
}

export function setOutput(value: unknown): void {
  const span = trace.getActiveSpan();
  if (span) span.setAttribute(Attr.OUTPUT_VALUE, toValue(value));
}

export interface SpanOptions {
  kind?: SpanKindValue;
  input?: unknown;
  attributes?: Record<string, string | number | boolean>;
}

/** Run `fn` inside a Beater span, recording status and re-raising errors. */
export async function span<T>(
  name: string,
  options: SpanOptions,
  fn: (span: Span) => Promise<T> | T,
): Promise<T> {
  const tracer = getTracer();
  return tracer.startActiveSpan(name, async (current) => {
    applyCommon(current, options.kind ?? SpanKind.AGENT_STEP);
    if (options.input !== undefined) current.setAttribute(Attr.INPUT_VALUE, toValue(options.input));
    for (const [key, val] of Object.entries(options.attributes ?? {})) {
      current.setAttribute(key, val);
    }
    try {
      const result = await fn(current);
      current.setStatus({ code: SpanStatusCode.OK });
      return result;
    } catch (err) {
      current.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
      current.recordException(err as Error);
      throw err;
    } finally {
      current.end();
    }
  });
}

export interface ObserveOptions {
  name?: string;
  kind?: SpanKindValue;
  captureInput?: boolean;
  captureOutput?: boolean;
}

/**
 * Wrap a function so each call becomes a Beater span. Works for sync and async
 * functions (the returned function is always async-safe).
 *
 *   const call = observe(rawCall, { kind: SpanKind.LLM_CALL });
 */
export function observe<A extends unknown[], R>(
  fn: (...args: A) => Promise<R> | R,
  options: ObserveOptions = {},
): (...args: A) => Promise<R> {
  const name = options.name ?? fn.name ?? "anonymous";
  const kind = options.kind ?? SpanKind.AGENT_STEP;
  const captureInput = options.captureInput ?? true;
  const captureOutput = options.captureOutput ?? true;

  return async (...args: A): Promise<R> => {
    const tracer = getTracer();
    return tracer.startActiveSpan(name, async (current) => {
      applyCommon(current, kind);
      if (captureInput && args.length > 0) current.setAttribute(Attr.INPUT_VALUE, toValue(args));
      try {
        const result = await fn(...args);
        if (captureOutput && result !== undefined) {
          current.setAttribute(Attr.OUTPUT_VALUE, toValue(result));
        }
        current.setStatus({ code: SpanStatusCode.OK });
        return result;
      } catch (err) {
        current.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
        current.recordException(err as Error);
        throw err;
      } finally {
        current.end();
      }
    });
  };
}

export { context };
export const _internal = { toValue, applyCommon };

File: sdks/python/beater/observe.py
"""The ``@observe`` decorator and ``span`` context manager.

These are the ergonomic entry points for tracing arbitrary code as Beater spans.
"""

from __future__ import annotations

import functools
import inspect
import itertools
import json
from contextlib import contextmanager
from typing import Any, Callable, Iterator, Optional

from opentelemetry.trace import Status, StatusCode

from .semconv import Attr, SpanKind
from .tracing import get_config, get_tracer

_seq = itertools.count(1)


def _to_value(obj: Any) -> str:
    if isinstance(obj, str):
        return obj
    try:
        return json.dumps(obj, default=str)
    except (TypeError, ValueError):
        return str(obj)


def _apply_common(span, kind: str) -> None:
    span.set_attribute(Attr.SPAN_KIND, kind)
    span.set_attribute(Attr.SEQ, next(_seq))
    config = get_config()
    if config and config.release_id:
        span.set_attribute(Attr.RELEASE_ID, config.release_id)


def set_input(value: Any) -> None:
    """Attach an input payload to the current span."""
    from opentelemetry import trace as _trace

    span = _trace.get_current_span()
    if span is not None:
        span.set_attribute(Attr.INPUT_VALUE, _to_value(value))


def set_output(value: Any) -> None:
    """Attach an output payload to the current span."""
    from opentelemetry import trace as _trace

    span = _trace.get_current_span()
    if span is not None:
        span.set_attribute(Attr.OUTPUT_VALUE, _to_value(value))


@contextmanager
def span(
    name: str,
    *,
    kind: str = SpanKind.AGENT_STEP,
    input: Any = None,
    attributes: Optional[dict] = None,
) -> Iterator[Any]:
    """Open a Beater span as a context manager.

    >>> with beater.span("retrieve", kind=SpanKind.RETRIEVAL_QUERY) as s:
    ...     ...
    """
    tracer = get_tracer()
    with tracer.start_as_current_span(name) as current:
        _apply_common(current, kind)
        if input is not None:
            current.set_attribute(Attr.INPUT_VALUE, _to_value(input))
        for key, val in (attributes or {}).items():
            current.set_attribute(key, val)
        try:
            yield current
            current.set_status(Status(StatusCode.OK))
        except Exception as exc:  # noqa: BLE001 - re-raised below
            current.set_status(Status(StatusCode.ERROR, str(exc)))
            current.record_exception(exc)
            raise


def observe(
    _func: Optional[Callable] = None,
    *,
    name: Optional[str] = None,
    kind: str = SpanKind.AGENT_STEP,
    capture_input: bool = True,
    capture_output: bool = True,
) -> Callable:
    """Decorate a function so each call becomes a Beater span.

    Works on sync and async functions. Captures arguments and the return value
    as ``input.value`` / ``output.value`` unless disabled.

    >>> @beater.observe(kind=SpanKind.LLM_CALL)
    ... def call_model(prompt): ...
    """

    def decorate(func: Callable) -> Callable:
        span_name = name or func.__qualname__

        def _record_input(current, args, kwargs):
            if capture_input:
                payload = {"args": args, "kwargs": kwargs} if (args or kwargs) else None
                if payload is not None:
                    current.set_attribute(Attr.INPUT_VALUE, _to_value(payload))

        if inspect.iscoroutinefunction(func):

            @functools.wraps(func)
            async def async_wrapper(*args, **kwargs):
                tracer = get_tracer()
                with tracer.start_as_current_span(span_name) as current:
                    _apply_common(current, kind)
                    _record_input(current, args, kwargs)
                    try:
                        result = await func(*args, **kwargs)
                        if capture_output and result is not None:
                            current.set_attribute(Attr.OUTPUT_VALUE, _to_value(result))
                        current.set_status(Status(StatusCode.OK))
                        return result
                    except Exception as exc:  # noqa: BLE001
                        current.set_status(Status(StatusCode.ERROR, str(exc)))
                        current.record_exception(exc)
                        raise

            return async_wrapper

        @functools.wraps(func)
        def sync_wrapper(*args, **kwargs):
            tracer = get_tracer()
            with tracer.start_as_current_span(span_name) as current:
                _apply_common(current, kind)
                _record_input(current, args, kwargs)
                try:
                    result = func(*args, **kwargs)
                    if capture_output and result is not None:
                        current.set_attribute(Attr.OUTPUT_VALUE, _to_value(result))
                    current.set_status(Status(StatusCode.OK))
                    return result
                except Exception as exc:  # noqa: BLE001
                    current.set_status(Status(StatusCode.ERROR, str(exc)))
                    current.record_exception(exc)
                    raise

        return sync_wrapper

    if _func is not None:
        return decorate(_func)
    return decorate
